### PR TITLE
Alerting: Add support to distinguish Prometheus datasource subtypes (Mimir, Cortex and Vanilla Prometheus)

### LINKED ISSUE #45825: Feature Request: Grafana 8.0 Ruler API not accesible on same URL as Cortex datasource
### Discussed in https://github.com/grafana/grafana/discussions/43662

<div type='discussions-op-text'>

<sup>Originally posted by **kubicgruenfeld** June  8, 2021</sup>
**What would you like to be added**:

The possiblity to configure the ruler API on an other url in Prometheus datasources since they must not be the same.

**Why is this needed**:

To use ngalert e.g. in Grafana deployments in k8s clusters with Cortex the url of the Prometheus datasources would most likely use something like `http://query-frontend/...` and the ruler would be reachable with an other service like `http://ruler/...`</div>

### PR DESCRIPTION
**Pull request by konrad147**
**What this PR does / why we need it**:
This PR tries to utilize the Prometheus buildinfo endpoint to check if the Ruler API is supported.

**Which issue(s) this PR fixes**:
Fixes #45825

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by peterholmberg** [comment]
is this different from https://github.com/grafana/grafana/pull/37064 ?

**Comment by konrad147** [comment]
I closed the #37064 as we need to take a different approach regarding custom ruler URL

File: public/app/features/alerting/unified/api/buildInfo.test.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        expect(response.application).toBe(PromApplication.Prometheus);
+        expect(response.features.rulerApiEnabled).toBe(rulerApiEnabled);
+        expect(mocks.fetchRules).not.toHaveBeenCalled();
+        expect(mocks.fetchTestRulerRulesGroup).not.toHaveBeenCalled();
+      }
+    );
+  });
+
+  describe('When buildinfo returns 404 error', () => {
+    it('Should return cortex with ruler API disabled when prom rules works and ruler api returns not avaiable errors', async () => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/api/buildInfo.test.ts:74]: nit: small typo here `available`

**Review by gillesdemey — APPROVED** [review]
This is very impressive, Konrad! 🙌

File: pkg/services/ngalert/api/lotex_ruler_test.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/infra/log"
```
Comment (joeblubaugh) [inline pkg/services/ngalert/api/lotex_ruler_test.go:9]: the formatter won't enforce this, but our preference for import formatting looks like:
```go
import (
  // stdlib packages
 "context"
 "errors"

  // non-grafana packages
  "github.com/stretchr/testify/require"
  
  // packages in the local repository
  "github.com/grafana/grafana/pkg/models"
  "github.com/grafana/grafana/pkg/web"
)
```

do you have an editor setting that's leaving these empty lines in-between packages?

File: pkg/services/ngalert/api/lotex_ruler.go [hunk 3]
```diff
@@ -16,9 +16,32 @@ import (
 	"github.com/grafana/grafana/pkg/models"
 )
 
+const (
+	Prometheus = iota + 1
```
[the remark this answers is not in the document]
Comment (joeblubaugh) [reply pkg/services/ngalert/api/lotex_ruler.go:20]: I agree with this - the API should use constant strings for parameters. These values aren't safe to re-order - by inserting "Mimir" between Prometheus and Cortex, for example - so using a string constant is safer from rogue refactorings. There's no performance benefit to using integers; they have to be parsed from strings in `QueryInt()`

[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/api/lotex_ruler.go]: Good spot, my initial thinking was to obfuscate Mimir a bit with the numbers + cloud wording in-code.

Fixed.

File: pkg/services/ngalert/api/lotex_ruler.go [hunk 4]
```diff
@@ -16,9 +16,32 @@ import (
...
 var dsTypeToRulerPrefix = map[string]string{
-	"prometheus": "/rules",
-	"loki":       "/api/prom/rules",
+	PrometheusDatasourceType: prometheusPrefix,
+	LokiDatasourceType:       lokiPrefix,
+}
+
+var subtypeToPrefix = map[string]string{
+	Prometheus: prometheusPrefix,
+	Cortex:     prometheusPrefix,
```
Comment (gotjosh) [inline pkg/services/ngalert/api/lotex_ruler.go:43]: Prometheus and Cortex share a path for the configuration API (and so does Mimir). However, there's more than meets the eye here (see more info in the original issue).

This complexity is meant to be tackled in #46771

**Comment by gillesdemey** [comment]
Added some additional visualisation to the verbose message returned by the Prometheus datasource to indicate if the datasource is either a `Prometheus`, `Mimir` or `Cortex` datasource and whether the ruler API is enabled.
image

**Review by gotjosh — APPROVED** [review]
LGTM, but I'm the author of some of these changes.

Would be good to get at least another +1 from non-committing reviewers.

**Comment by joeblubaugh** [comment]
I don't feel qualified to review the TS, but I have a meta-review:
Could this have been broken up into smaller PRs? How many do we feel would be right? Could we have used configuration to let us deliver this in incremental pieces with tests even if the full feature set wasn't ready for use?

**Comment by gotjosh** [comment]
@peterholmberg or @aocenas can I get a quick re-review of the JS changes? This has already been seen by 3+ people and both Giles and I manually tested this extensively yesterday.

> Could this have been broken up into smaller PRs? How many do we feel would be right? Could we have used configuration to let us deliver this in incremental pieces with tests even if the full feature set wasn't ready for use?

I'll let the author of the PR drive this question, but if I had to guess my impression is that splitting this into multiple PRs would have made it much harder to get the right context all in one place. I believe he did make a split of one part with https://github.com/grafana/grafana/pull/46960

**Review by gillesdemey — APPROVED** [review]

File: public/app/plugins/datasource/prometheus/datasource.tsx [hunk 5]
```diff
@@ -811,6 +824,71 @@ export class PrometheusDatasource
...
+          display: 'grid',
+          gridTemplateColumns: 'max-content max-content',
+          rowGap: '0.5rem',
+          columnGap: '2rem',
+          marginTop: '1rem',
+        }}
+      >
+        <div>Type</div>
+        <div>{applicationSubType}</div>
+        {/* Prometheus does not have a Ruler API – so skip for that type */}
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/datasource.tsx:881]: Don't we want to tell Prometheus users that some functionality won't be available for them in that case?

**Review by aocenas — APPROVED** [review]
Still have a nitpick comment about the build message but overall code looks good.

File: public/app/plugins/datasource/prometheus/datasource.tsx (hunk 5 above)
Comment (gillesdemey) [reply public/app/plugins/datasource/prometheus/datasource.tsx:881]: Good suggestion, I think we should try to include that. We've seen some instances where customers are confused about not being able to manage rules with their vanilla Prometheus datasource.

Comment (gillesdemey) [reply public/app/plugins/datasource/prometheus/datasource.tsx:881]: Added an "unsupported" badge and a tooltip to clarify to the user that we do not support rule management with Prometheus.

image

Comment (konrad147) [reply public/app/plugins/datasource/prometheus/datasource.tsx]: Awesome! 🚀

**Comment by konrad147** [comment]
> I don't feel qualified to review the TS, but I have a meta-review: Could this have been broken up into smaller PRs? How many do we feel would be right? Could we have used configuration to let us deliver this in incremental pieces with tests even if the full feature set wasn't ready for use?

Good point. I feel I could have put fetching the build info information into a separate PR earlier. Although, it wasn't so obvious at the beginning if the current approach is the correct way to go.

**Comment by madushan1000** [comment]
Is this change included in grafana 8.4.6? or only in 8.5.0 beta?

**Comment by sakjur** [comment]
@madushan1000 It's only in the 8.5.0 beta.

**Comment by hpl002** [comment]
Apologies for the necrobump, but i spent a few hours scratching my head at this so its a worthy mention.
Would it be possible to add a similar warning in the  _alerting_ panel for alerts that use a prometheus style datasource? 

I have a collection of provisioned prometheus alerts that are successfully added and trigger as expected, but i keep getting HTTP404 whenever the frontend tries to do some lookup via the ruler API. I'm presuming that if fails due to prometheus not being supported.



Err
```shell 
grafana-58b59b949d-4p9df grafana logger=context userId=2 orgId=1 uname=hermanp@stacc.com t=2023-02-28T13:45:22.273117398Z level=error msg="no rule groups found\n" error="no rule groups found\n" remote_addr=82.134.13.181 traceID=
grafana-58b59b949d-4p9df grafana logger=context userId=2 orgId=1 uname=hermanp@stacc.com t=2023-02-28T13:45:22.273179558Z level=info msg="Request Completed" method=GET path=/api/ruler/P8E80F9AEF21F6940/api/v1/rules status=404 remote_addr=82.134.13.181 time_ms=9 duration=9.563578ms size=49 referer=https://grafana.hermantemp.stacc.dev/alerting/list handler=/api/ruler/:DatasourceUID/api/v1/rules
```

**Pull request merged**